In [3]:
import os

raw_path = r"C:\Users\pavansai\OneDrive\Desktop\RetailPulse\data\raw"
os.listdir(raw_path)


['.ipynb_checkpoints', 'online_retail.xlsx']

In [4]:
import pandas as pd

file_path = r"C:\Users\pavansai\OneDrive\Desktop\RetailPulse\data\raw\online_retail.xlsx"

df = pd.read_excel(file_path, engine="openpyxl")

df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [5]:
df.shape

(541909, 8)

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 40.0+ MB


In [7]:
df.isnull().sum()

InvoiceNo           0
StockCode           0
Description      1454
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     135080
Country             0
dtype: int64

In [8]:
df_clean = df.copy()

# Remove rows where CustomerID is missing
df_clean = df_clean.dropna(subset=["CustomerID"])

# Remove cancelled/returned orders
df_clean = df_clean[~df_clean["InvoiceNo"].astype(str).str.startswith("C")]

# Remove invalid quantity and price
df_clean = df_clean[df_clean["Quantity"] > 0]
df_clean = df_clean[df_clean["UnitPrice"] > 0]

# Convert CustomerID from decimal to integer
df_clean["CustomerID"] = df_clean["CustomerID"].astype(int)

# Create revenue column
df_clean["Revenue"] = df_clean["Quantity"] * df_clean["UnitPrice"]

df_clean.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country,Revenue
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850,United Kingdom,15.30
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850,United Kingdom,22.00
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850,United Kingdom,20.34


In [9]:
df_clean.shape

(397884, 9)

In [10]:
df_clean.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
Revenue        0
dtype: int64

In [11]:
df_clean.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID,Revenue
count,397884.000000,397884,397884.000000,397884.000000,397884.000000
mean,12.988238,2011-07-10 23:41:23.511023,3.116488,15294.423453,22.397000
min,1.000000,2010-12-01 08:26:00,0.001000,12346.000000,0.001000
25%,2.000000,2011-04-07 11:12:00,1.250000,13969.000000,4.680000
50%,6.000000,2011-07-31 14:39:00,1.950000,15159.000000,11.800000
75%,12.000000,2011-10-20 14:33:00,3.750000,16795.000000,19.800000
max,80995.000000,2011-12-09 12:50:00,8142.750000,18287.000000,168469.600000
std,179.331775,NaN,22.097877,1713.141560,309.071041


In [13]:
df_clean.to_csv(
    r"C:\Users\pavansai\OneDrive\Desktop\RetailPulse\data\processed\cleaned_online_retail.csv",
    index=False
)


In [14]:
import os

processed_path = r"C:\Users\pavansai\OneDrive\Desktop\RetailPulse\data\processed"

if os.path.isfile(processed_path):
    os.remove(processed_path)

os.makedirs(processed_path, exist_ok=True)

df_clean.to_csv(
    r"C:\Users\pavansai\OneDrive\Desktop\RetailPulse\data\processed\cleaned_online_retail.csv",
    index=False
)

In [15]:
os.listdir(processed_path)

['cleaned_online_retail.csv']

In [16]:
['cleaned_online_retail.csv']

['cleaned_online_retail.csv']

## Data Cleaning Summary

Original dataset shape: 541909 rows and 8 columns.

Cleaning steps performed:
1. Removed rows with missing CustomerID.
2. Removed cancelled invoices starting with C.
3. Removed rows where Quantity <= 0.
4. Removed rows where UnitPrice <= 0.
5. Converted CustomerID into integer format.
6. Created Revenue column using Quantity * UnitPrice.
7. Saved cleaned data as cleaned_online_retail.csv.

Final cleaned dataset shape: 397884 rows and 9 columns.